control/controlhttp/constants.go

```diff
@@ -18,3 +28,51 @@ const (
 	// to do the protocol switch is located.
 	serverUpgradePath = "/ts2021"
 )
+
+// Dialer contains configuration on how to dial the Tailscale control server.
+type Dialer struct {
+	// Host is the host to connect to.
+	//
+	// This field is required.
+	Host string
+
+	// HTTPPort is the port number to use when making a HTTP connection.
+	//
+	// This field is required.
+	HTTPPort string
+
+	// HTTPSPort is the port number to use when making a HTTPS connection.
+	//
+	// This field is required.
+	HTTPSPort string
+
+	// MachineKey contains the current machine's private key.
+	//
+	// This field is required.
+	MachineKey key.MachinePrivate
+
+	// ControlKey contains the expected public key for the control server.
+	//
+	// This field is required.
+	ControlKey key.MachinePublic
+
+	// ProtocolVersion is the expected protocol version to negotiate.
+	//
+	// This field is required.
+	ProtocolVersion uint16
+
+	// Dialer is the dialer used to make outbound connections.
+	//
+	// This field is required.
+	Dialer dnscache.DialContextFunc
```

Comment:
why? net.Dial seems like a fine default?

Reply:
For this (and the other cases) I was replicating the prior behaviour; happy to change to a sensible default.

Reply:
yeah, let's. I think it's just those 3 fields that have pretty obvious defaults.

Reply:
Done; added defaults and updated comments.